src/ocean/task/TaskPool.d

```diff
@@ -361,3 +371,44 @@ unittest
     theScheduler.schedule(new MainTask);
     theScheduler.eventLoop();
 }
+
+unittest
+{
+    // Test that 'awaitRunningTasks()' cannot be called from a task that itself
+    // belongs to the pool.
+
+    static class AwaitTask : Task
+    {
+        void delegate () non_throwing_dg;
+        void delegate () throwing_dg;
+
+        public void copyArguments ( void delegate () non_throwing_dg,
+            void delegate () throwing_dg )
+        {
+            this.non_throwing_dg = non_throwing_dg;
+            this.throwing_dg = throwing_dg;
+        }
+
+        override public void run ( )
+        {
+            this.non_throwing_dg();
```

Comment:
>Why this no-op delegate call is needed at all? Looks confusing.

OK, I'll get rid of it.

>Also can copy pool reference instead of delegate and call `pool.awaitRunningTasks` directly from `run`

How can I copy it without running into the forward reference errors?
That is, if I do this:

```d
unittest
{
    // Test that 'awaitRunningTasks()' cannot be called from a task that itself
    // belongs to the pool.

    static class AwaitTask : Task
    { 
        private TaskPool!(AwaitTask) pool;

        public void copyArguments ( TaskPool!(AwaitTask) pool ) 
        { 
            this.pool = pool;
        } 

        override public void run ( ) 
        { 
            this.pool.awaitRunningTasks();
        } 
    } 

    auto pool = new TaskPool!(AwaitTask);
    initScheduler(SchedulerConfiguration.init);

    pool.start(pool);
}
```

it fails with:
```
Error: template instance ocean.core.Traits_tango.ParameterTupleOf!(copyArguments) forward reference of copyArguments
```